### DISCUSSION TITLE
Nrule - multiple rules to validate and add error message in single property

### DISCUSSION THREAD

**Discussion by pawanrai** [NRules/NRules#302] [General]
We just started using Nrules in C#.net. There are multiple complex checks in our system that we are creating separate Nrule classes for. But in Then/Action method of Nrule classes we are appending text to the errortext property of the object like below:

obj.ErrorText += "This is error No#1"

Now, whenever we are using single session to load rules and then facts and perform the rule engine check then in single object's errortext property in getting value n time where n is the number of classes representing different rules that were loaded in session. However, this error message is because the first rule class ran and identified this error.

However, when we load and execute rules in separate sessions and fire them separately it works perfectly fine.

Could anyone please suggest what is wrong with the implementation.

Thanks in Advance.

**Comment by snikolayev** [comment]
@pawanrai to be honest I don’t understand the explanation of the problem when you load rules into one session. Maybe you have a small repro scenario or can explain it differently?

**Reply by pawanrai** [reply]
Let me explain details using code. Here are the Nrule class definitions:

```
 public class FirstRule : Rule
    {
        public override void Define()
        {
		Customer cst = default;
           
            When()
   .Match<Customer>(() => cst, x => some conditional lambda here);

            Then()
              .Do(ctx => ApplyText(cst));
        }

        private static void ApplyText(Object abc)
        {
            abc.ErrorText += "This customer matches the first rule";
        }

    }

 public class SecondRule : Rule
    {
        public override void Define()
        {
		Customer cst = default;
           
            When()
   .Match<Customer>(() => cst, x => some conditional lambda here);

            Then()
              .Do(ctx => ApplyText(cst));
        }

        private static void ApplyText(Object abc)
        {
            abc.ErrorText += "This customer matches the Second rule";
        }

    }

```
And here is the module loading and applying rules and facts:
```
var repository = new RuleRepository();
repository.Load(x => x.From(typeof(FirstRule).Assembly));
repository.Load(x => x.From(typeof(SecondRule).Assembly));
var factory = repository.Compile();

var session = factory.CreateSession();

//Code to load list of customer objects in session goes here
session.InsertAll(customerList);
session.Fire();
```

Now, suppose the first customer object did match the first rule only and no other match is found. In this case the first customers **ErrorText** property has below value:

**_This customer matches the first ruleThis customer matches the first rule_**

Notice the errorText has the same value twice. Since we have 3 rules this value is getting added to **ErrorText** property thrice in same object that matches first rule.

However, in case we load the rules separately in different sessions, it works fine.

Hope it clarifies.

**Reply by pawanrai** [reply]
Just an updated. I used the Fluent Rule Filter mechanism stated at https://github.com/NRules/NRules/wiki/Agenda-Filters to make sure that if one rule makes changes to object then the other rule is not called again unless the properties part of the check are changed. So it worked.

However, now the problem is that though we are just loading the first rule class in the session still all three rules are being run.

**Comment by pawanrai** [comment]
I found the problem. All the rules were part of same assembly, hence adding a condition to only load specific rules worked...

Thanks snikolayev for looking into this
